executor/src/core.rs

```diff
@@ -143,54 +147,175 @@ where
                 // the second batch since there is no point in executing twice the same transactions
                 // (as the second execution attempt will always fail).
                 debug!("Duplicate batch {batch_digest}");
-                self.execution_indices.skip_batch(total_batches);
-                return Ok(());
+                return Ok(Vec::new());
             }
         };
 
+        // The consensus simply orders bytes, so we first need to deserialize the transaction.
+        // If the deserialization fail it is safe to ignore the transaction since all correct
+        // clients will do the same. Remember that a bad authority or client may input random
+        // bytes to the consensus.
+        let transactions = transactions
+            .into_iter()
+            .map(
+                |serialized| match bincode::deserialize::<State::Transaction>(&serialized) {
+                    Ok(x) => Ok(x),
+                    Err(e) => Err(SubscriberError::ClientExecutionError(format!(
+                        "Failed to deserialize transaction: {e}"
+                    ))),
+                },
+            )
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(transactions)
+    }
+}
+
+/// Executor that feeds transactions one by one to the execution state.
+pub struct SingleExecutor<State>
+where
+    State: SingleExecutionState,
+{
+    /// The (global) state to perform execution.
+    execution_state: Arc<State>,
+    /// The indices ensuring we do not execute twice the same transaction.
+    execution_indices: Mutex<ExecutionIndices>,
```

Comment:
This Mutex seems dangerous. Do we absolutely need it?

Reply:
It is unfortunate, I'm not sure if there's another way. The only reason it's here is because the `SingleExecutor` instance has to be `Sync` and `Send` so we can call its async methods, and it also cannot be mutable (the traits methods aren't, and it's shared through an `Arc`), so the internally mutable indices must be put in something. Unfortunately `RefCell` cannot be shared across threads, and the next option was the `Mutex`. It is held between async calls, so it's the future specific variant.

It's only ever used by this instance, though, and it's only called sequentially, so there's no risk of a deadlock I think, but it must have a negative performance impact.